### PR TITLE
Rework timed submissions, support block / force submit at deadline

### PR DESCRIPTION
**Pull request by adi-herwana-nus**
Adds a per-assessment **"Allow submissions after end date"** toggle. When disabled, students can no
longer create or edit submissions past the end date — any in-progress attempt is **force-submitted**
at the end date instead of merely being flagged late. Along the way this reworks the existing
timed-assessment force-submission onto one unified, more robust mechanism.

Four commits:

1. `feat(submissions): add flag to allow/block students from submissions past deadline` — the toggle, data model, and enforcement.
2. `feat(submissions): rework backend implementation for time-based force-submission` — unify deadline + time-limit force submission; scheduling; fail-safe.
3. `feat(submissions): adapt frontend controls for force-submitted submissions` — client force-submit timer, countdown banner.
4. `feat(submissions): notify users of attempt block in assessment UI` — closed-state button, instructor settings row, terminology.

---

## Background

The following query was used to gather existing data about timed submissions in production:
```sql
SELECT
  ci.host,
  c.id														                           AS course_id,
  s.assessment_id											                       AS assessment_id,
  s.id                                                       AS submission_id,
  a.time_limit                                               AS time_limit_minutes,
  EXTRACT(EPOCH FROM (s.submitted_at - s.created_at)) / 60.0 AS duration_minutes
FROM course_assessment_submissions s
JOIN course_assessments a ON a.id = s.assessment_id
JOIN course_assessment_tabs cat ON cat.id=a.tab_id
JOIN course_assessment_categories cac ON cac.id=cat.category_id
JOIN courses c ON c.id=cac.course_id
JOIN instances ci ON ci.id=c.instance_id
WHERE a.time_limit IS NOT NULL
  AND s.submitted_at IS NOT NULL
  AND c.id != 2304 -- Testing course
  ORDER BY a.time_limit ASC;
```

The current implementation of timed submission has two force-submit operations:
- a client-side auto submit that fires with per-second granularity, but is based on the clock of each client;
- a server-side auto submit that reaps the remaining submissions at `end_at + 5.minutes` (e.g. when client has closed the submission page, their client cannot force submit)

Data shows that both operations are regularly exercised. Outliers are likely due to instructor intervention (manual unsubmitting, deadline changes, force submission job failure)
finalize_delta_histogram

Also worth noting, exactly 4 timed assessments have a deadline of exactly 6 hours, and 2 have a deadline exceeding 6 hours.

---

Also worth noting, exactly 4 timed assessments have a deadline of exactly 6 hours, and 2 have a deadline exceeding 6 hours.

---

## Key changes

Beyond adding end-date enforcement, this PR changes the existing timed-submission process as follows. A
"time limit" and an "end date" are now just two sources of the same force-submit time, so all of this
applies to both.

1. **Remaining time is forwarded to the client, not the time limit.** The client used to compute the
   auto-submit moment from `time_limit` and the submission's `created_at`, measured against *each
   client's own clock*. The server now sends the remaining duration directly (the earlier of time-limit
   expiry and the end date), which the client anchors to its local clock once at load. A skewed client
   clock can no longer fire the auto-submit early (cutting a student short) or late, and the entry
   warning dialog and the countdown banner now show the same value.

2. **The client auto-submit timer is torn down when the component unmounts.** The effect that arms the
   timer now returns its cleanup, so leaving the submission page clears the interval. Previously it
   leaked and kept firing after unmount, racing with the timer re-armed on the next page — which could
   dispatch several concurrent `finalise` auto-submit requests. This fixes that race.

3. **Long-running timed submissions (> 6h) are force-submitted via the hourly sweep, not at create/update.**
   A force-submit job is enqueued eagerly only when its time runs out within
   `FORCE_SUBMIT_SCHEDULING_HORIZON` (6h). Anything further out is picked up by the hourly
   `ScheduleExpiringSubmissionsJob` sweep once it comes within the horizon. This bounds how long a job
   sits in the queue (and thus how stale its serialized arguments can get) and how far ahead we commit;
   the sweep also reschedules on end-date / personal-time changes and recovers lost jobs. (Per Background,
   only ~2 existing assessments exceed 6h.)

4. **Fail-safe: if the creator resumes past the end date, force-submit directly before rendering.** If the
   submission's *own creator* opens the edit page for an attempting submission that is already overdue
   (its scheduled job never ran), it is force-submitted *before* any edit data is returned, so the page
   never opens as a resumable past-due attempt.
   - **Synchronous:** yes. A `before_action` calls `force_submit!` — `update!('finalise' => 'true')`
     inside a transaction — in the request thread; the response blocks until the finalise's synchronous
     work completes (auto-grading is enqueued asynchronously).
   - **On rollback:** `update!` raises, the transaction rolls back (the submission stays `attempting`,
     untouched), and — the edit path has no rescue — the error surfaces as a 500 rather than rendering a
     resumable attempt. Fail-loud by design: the student is never shown a past-due resumable page.

5. **Any unsubmitted submission is fully exempt from the force-submission workflow.** A submission
   unsubmitted by staff (a permitted redo) is removed from *every* force-submission path — no scheduling,
   no sweep, no fail-safe — and editing is unblocked, until it is finalised again.

6. **Other user-facing changes.** The `LATE` badge is never shown when late submissions are disallowed
   (nothing can be late by construction); a closed assessment shows a disabled "Attempt" button with an
   explanatory tooltip; and instructors see the toggle's value as a settings row (when an end date is set).

---

## What changed

**Data model** (`AddLateSubmissionSupport`, one migration)
- `course_assessments.is_late_submission_allowed` (boolean, default `true` — existing behaviour preserved).
- `course_assessment_submissions.unsubmitted_at` — marks a staff-permitted redo, exempt from force
  submission (see Design decisions).
- `course_assessment_submissions.force_submit_scheduled_at` — the sweep's dedup/reschedule marker.

**Model** (`Course::Assessment`, `Course::Assessment::Submission`)
- `submission_deadline_for` / `submission_deadline_passed_for?` (strict) / `editing_deadline_passed_for?`
  (with grace) — the enforcement predicates, personalised-timeline aware.
- `force_submit_at` = **the earlier of time-limit expiry and effective end date** — the single source of
  truth, nil when never force-submitted or when unsubmitted.
- `force_submit_job_at`, `force_submit_overdue?`, `force_submit!`, `schedule_force_submission`,
  `enqueue_force_submit_job`, `force_submit_job_scheduled?`.
- `Course::Assessment.to_force_submit` scope.

**Jobs**
- `ForceSubmitTimedSubmissionJob` — now finalises via `submission.force_submit!` and **re-checks live
  state** (`attempting?`, `force_submit_at > now`) before acting; no token argument.
- `ScheduleExpiringSubmissionsJob` (hourly cron) — schedules precise future-dated force-submit jobs for
  submissions whose force-submit time is imminent and not yet scheduled; reschedules on end-date /
  personal-time change; recovers lost jobs. Eager-loads the timeline records `time_for` reads, to avoid
  an N+1 across submissions.

**Enforcement**
- **Ability (`define_student_assessment_permissions`): `cannot :update`/`:submit_answer` past the edit grace.**
- `SubmissionsController`: `check_submission_deadline!` (edit, grace), a strict end-date check in
  `authorize_assessment!` (create), and `force_submit_if_overdue!` before authorization (edit fail-safe).
- `LATE` flag: never flagged when late submissions are disallowed; comparison fixed to Time-vs-Time.

**Client**
- **Server sends `forceSubmitRemainingTime`; the timer, countdown banner, and entry warning dialog all
  anchor it to the local clock.**
- Edit page arms `setTimerForForceSubmission` **(cleanup on unmount)** and shows the countdown banner
  (hidden > 24h). `WarningDialog` now uses the unified remaining time (so it also fires for deadline-only
  assessments, not just timed ones).
- Assessment form: the toggle (disabled without an end date / on Koditsu).
- Closed state: a disabled "Attempt" button with a tooltip; an instructor settings row (shown only when
  an end date is set).

**i18n** — new keys hand-added to `locales/{en,ko,zh}.json` and Rails `config/locales/{en,ko,zh}`.

---

## Design decisions

1. **Force submission *is* the enforcement**, not just a disabled form. Once a submission leaves
   `attempting`, existing ability rules make it read-only — so finalising at the end date enforces "no
   edits after end date" for free, and preserves the student's in-progress answers (the client submits
   them with the finalise). Mirrors the timed-assessment dual mechanism (client timer + server backstop).

2. **`:attempt` ability left intact.** Blocking it also revokes `read_material`, so a past-end-date
   student could no longer even *view* the assessment or their own submission. Instead edits are blocked
   by `cannot :update` (students only) and creation by an explicit controller check.

3. **Unified time-limit + end-date force submission** behind `force_submit_at`. One code path, one
   source of truth. Also the natural extension point for future *per-student time limits* — it becomes a
   change to `force_submit_at` alone.

4. **Grace for editing, strict for creation.** Editing an existing attempt is allowed until
   `end_date + FORCE_SUBMIT_DELAY` (5 min): the client fires its own force-submit `finalise` just after
   the end date (carrying the latest answers), and a hard cut-off would 403 that request and lose work.
   The histogram in Background bears this out — a large share of timed submissions finalise in exactly
   this window, right as the client auto-submit fires. Creating a *new* submission is strict (no
   in-progress work to preserve).

5. **No dedup token; a per-submission marker instead.** The token's "invalidate a stale job" role is
   already covered by the job's `force_submit_at > now` re-check. The real gap was *rescheduling* after
   an end-date change, which a per-assessment token can't do (and it misses per-student personal-time
   changes). `force_submit_scheduled_at` records the time a job was scheduled for; the sweep reschedules
   when it no longer matches, and dedups to zero otherwise — important for timed exams where every
   submission is attempting near the deadline.

6. **Cron pre-schedules precise jobs; it is not an "overdue sweep".** Firing a future-dated job at the
   exact moment preserves minute-granularity; a sweep that force-submits inline every N minutes would
   lose it. The marker makes a wide look-ahead window and an infrequent (hourly) run safe.

7. **Unsubmitted submissions are exempt** from enforcement and force submission (staff explicitly
   permitted the redo). Tracked with a dedicated `unsubmitted_at` column — **not** `submitted_at`, which
   `unsubmit` nils, and which `learning_rate_concern` / `counts_concern` read as "was submitted" *without*
   a workflow filter, so retaining it would skew statistics.

8. **The edit-page fail-safe is creator-only.** An instructor or admin opening a student's in-progress
   attempt must not force-submit it on the student's behalf.

9. **Remaining duration, not an absolute timestamp, is sent to the client** and anchored to the local
   clock at the timer/banner (a side-effect at the leaf, not in the reducer). A skewed client clock then
   can't fire the force-submit early (cutting a student short) or late.

10. **Terminology: "end date", not "deadline"** — matches the existing *End At* field.

### Pre-existing bugs fixed along the way
- `late` compared a `String` (`submitted_at.iso8601`) to a `TimeWithZone` (`end_at`) — worked only via
  opaque coercion; now a clean Time comparison.
- The `TimeLimitBanner.seconds` message shared `minutesSeconds`'s message id (so the sub-minute
  countdown resolved the wrong string); given its own id.

---

## Testing

- **Ruby:** model, ability, controller, and job specs for the deadline predicates, strict-vs-grace
  enforcement, the edit fail-safe (creator-only + unsubmitted exemption), the scheduler (schedule /
  reschedule-on-change / dedup / lost-job recovery / unsubmitted exclusion), the force-submit job's
  re-check, and the `late` flag. Regression suites green.
- **Client:** jest for `getForceSubmitRemainingTime`, the force-submit timer (fires on expiry; **disarmed
  on navigate-away**), and the closed-state button (disabled, no link, tooltip on hover). tsc + eslint clean.
- **Rails 8 compatibility** verified (the branch rebased across a 7.2→8.0 upgrade).

## Notes / follow-ups
- `FORCE_SUBMIT_SCHEDULING_HORIZON` (6h), `SCHEDULE_WINDOW` (horizon + 30m), and the hourly cron are
  tunable dials.
- The closed "Attempt" button uses the strict end date (appears exactly at the end date), consistent
  with strict creation.
- The token-based **closing/opening reminder** jobs still schedule months-ahead `wait_until` jobs; the
  sweep-plus-short-horizon pattern established here is the argument for migrating those too (separate PR).

### CODE REVIEW

File: app/controllers/course/assessment/submission/answer/answers_controller.rb [hunk 1]
```diff
@@ -7,6 +7,7 @@ class Course::Assessment::Submission::Answer::AnswersController <
 
   before_action :authorize_submission!
   before_action :check_password, only: [:update]
+  before_action :check_submission_deadline!, only: [:update, :submit_answer]
```
[the remark this answers is not in the document]
Comment (adi-herwana-nus) [reply app/controllers/course/assessment/submission/answer/answers_controller.rb:10]: As long as they are blocked, this is not an issue

File: db/migrate/20260804222525_add_late_submission_support.rb [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    # When false, students may not create or edit submissions past the assessment deadline; any
+    # attempting submission is force-submitted at the deadline instead. Defaults to true to preserve
+    # the existing behaviour (late submissions permitted, merely flagged) for all existing records.
+    add_column :course_assessments, :is_late_submission_allowed, :boolean, default: true, null: false
+
+    # Set when a submission is unsubmitted by staff, and cleared when it is (re-)finalised. A present
+    # value marks a submission that a student is redoing with explicit staff permission; such a
+    # submission is exempt from deadline enforcement and force submission. Distinct from submitted_at,
+    # which unsubmit nils and which statistics rely on to mean "was submitted".
+    add_column :course_assessment_submissions, :unsubmitted_at, :datetime
```
[the remark this answers is not in the document]
Comment (adi-herwana-nus) [reply db/migrate/20260804222525_add_late_submission_support.rb:13]: Not a major concern, un-submission is meant to be relatively rare event.

File: app/models/course/assessment/assessment_ability.rb [hunk 3]
```diff
@@ -97,6 +97,23 @@ def define_student_assessment_permissions
...
+  # never subject to this. Reading the submission is unaffected — only edits are blocked.
+  def disallow_late_assessment_submission
+    cannot :update, Course::Assessment::Submission do |submission|
+      submission.editing_deadline_passed_for?(course_user)
+    end
+    cannot [:update, :submit_answer], Course::Assessment::Answer do |answer|
+      answer.submission.editing_deadline_passed_for?(course_user)
+    end
+    cannot :destroy_attachment, Course::Assessment::Answer::TextResponse do |text_response|
+      text_response.answer.submission.editing_deadline_passed_for?(course_user)
```
[the remark this answers is not in the document]
Comment (adi-herwana-nus) [reply app/models/course/assessment/assessment_ability.rb:115]: We will address in the next PR that covers scribing questions.
